frontend/sozisel-app/src/components/PresenterSession/EventsTimline/EventsTimeline.tsx

```diff
@@ -0,0 +1,220 @@
+import CalendarTodayIcon from "@material-ui/icons/CalendarToday";
+import { Event } from "../../../model/Template";
+import { Participant } from "../../../hooks/useLiveSessionParticipation";
+import PlayCircleFilledIcon from "@material-ui/icons/PlayCircleFilled";
+import QuizDetails from "./QuizDetails/QuizDetails";
+import useCountdownTimer from "../../../hooks/useCountdownTimer";
+import { useHistory } from "react-router";
+import { useTranslation } from "react-i18next";
+
+interface LaunchedEvent {
+  id: string;
+  insertedAt: string;
+}
+
+interface EventsTimelineProps {
+  events: Event[];
+  launchedEvents: LaunchedEvent[];
+  sessionId: string;
+  participants: Participant[];
+}
+
+interface TimerProps {
+  onFinishCallback: () => void;
+  startValue: number;
+}
+
+const Timer = ({ onFinishCallback, startValue }: TimerProps): ReactElement => {
+  const countdownTimer = useCountdownTimer({
+    startValue,
+    onFinishCallback,
+  });
+
+  return <p className="Timer">{countdownTimer}</p>;
+};
+
+const getRandomParticipants = (
+  participants: Participant[],
+  targetPercentageOfParticipants: number
+): string[] => {
+  return participants
+    .sort(() => 0.5 - Math.random())
+    .slice(
+      0,
+      Math.floor((targetPercentageOfParticipants / 100) * participants.length)
+    )
+    .map((participant) => participant.id);
+};
+
+export default function EventsTimeline({
+  events,
+  launchedEvents,
+  sessionId,
+  participants,
+}: EventsTimelineProps): ReactElement {
+  const { t } = useTranslation("common");
+  const [activeEventIdx, setActiveEventIdx] = useState<number>(0);
+  const [activeEventId, setActiveEventId] = useState<string>("");
+  const [activeEventCurrentSec, setActiveEventCurrentSec] = useState<number>(0);
+  const history = useHistory();
+
+  const [endSessionMutation, { error: endSessionError }] =
+    useEndSessionMutation();
+
+  const onEndSession = async () => {
+    await endSessionMutation({ variables: { id: sessionId } });
+    if (!endSessionError) {
+      history.push(`result/summary`);
+    }
+  };
+
+  useEffect(() => {
+    const lastEventIdx = launchedEvents.length - 1;
+    if (lastEventIdx >= 0) {
+      const lastEvent = launchedEvents[launchedEvents.length - 1];
+      const lastEventStartTime = moment.tz(
+        lastEvent.insertedAt,
+        "Europe/Warsaw"
+      );
+      const currentTime = Math.floor(
+        moment
+          .duration(moment.tz("Europe/Warsaw").diff(lastEventStartTime))
+          .asSeconds()
+      );
+      if (currentTime < events[lastEventIdx].eventData.durationTimeSec) {
+        setActiveEventIdx(lastEventIdx);
+        setActiveEventId(lastEvent.id);
+        setActiveEventCurrentSec(
+          events[lastEventIdx].eventData.durationTimeSec - currentTime
+        );
+      } else {
+        setActiveEventIdx(lastEventIdx + 1);
+      }
+    }
+  }, [activeEventIdx, events, launchedEvents]);
+
+  const [launchEventMutation, { error: launchEventError }] =
```

Comment:
Keep those mutations and other hooks together (hooks first and then callbacks), you are mixing some random functions with actual hooks.